### ISSUE TITLE
Can't Train a Normal CNN.

### ISSUE THREAD

**Issue by adithya-tp** [mlpack/mlpack#2146]
#### Issue description

When I try to train the following cnn, it only runs when I include the line 

`arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");`

However, if I comment out this line, the code no longer runs (both cases compile. but in this second case doesn't run - "Segmentation Fault"). It doesn't make sense to me why this happens.

#### Your environment

 * version of mlpack: 3.2.2
 * operating system: ubuntu 18.04
 * compiler: g++ 7.4.0
 * version of dependencies (Boost/Armadillo): libboost-1.65.1, armadillo-9.800.3

#### Steps to reproduce

 Here's the code to reproduce the issue: 

    #include <mlpack/core.hpp>
    #include <mlpack/core/data/split_data.hpp>
    #include <mlpack/methods/ann/layer/layer.hpp>
    #include <mlpack/methods/ann/ffn.hpp>
    #include <ensmallen.hpp>

    using namespace mlpack;
    using namespace mlpack::ann;
    using namespace arma;
    using namespace std;
    using namespace ens;

    int main()
    {
        double RATIO = 0.1;
        int CYCLES = 10;
        const int ITERATIONS_PER_CYCLE = 1;
        double STEP_SIZE = 10e-3;
        int BATCH_SIZE = 1;

        // Reading in image.
        cout << "Reading data ..." << endl;
        arma::Mat<unsigned char> matrix;
        data::ImageInfo info;
        data::Load("saitama.png", matrix, info, false, true);

        FFN<NegativeLogLikelihood<>, RandomInitialization> model;

        model.Add<Convolution<> >(info.Channels(), 6, 5, 5, 1, 1, 0, 0, info.Width(), info.Height());
        model.Add<ReLULayer<> >();
        model.Add<MaxPooling<> >(2, 2, 2, 2, true);
        model.Add<Linear<> >(62*62*6, 2);
        model.Add<LogSoftMax<> >();
        cout << "Training ..." << endl;

        // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
        SGD<AdamUpdate> optimizer(
          STEP_SIZE,
          BATCH_SIZE,
          ITERATIONS_PER_CYCLE,
          1e-8,
          true,
          AdamUpdate(1e-8, 0.9, 0.999));

        arma::mat trainY;
        trainY << 1;// << 0;
        arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");
        trainY.print("Train label: \n");
        cout << trainY.n_rows << " " << trainY.n_cols;

        for (int i = 1; i <= CYCLES; i++)
        {
          model.Train(arma::conv_to<arma::mat>::from(matrix), trainY, optimizer, PrintLoss(), ProgressBar());
        }
        return 0;
    }

(The image size I used is 128 x 128).

compiled using: 
```g++ cnn.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp -DHAS_STB```
#### Expected behavior

When the line is commented out, the network should still train. The code should run irrespective of whether the line in question is commented out or not.

#### Actual behavior

The code runs only when the line is NOT commented out.

**Comment by zoq** [comment]
Same behaviour if you convert the matrix before you pass it?

```c++
arma::mat input = arma::conv_to<arma::mat>::from(matrix);
model.Train(input, trainY, optimizer, PrintLoss(), ProgressBar());
```

**Comment by adithya-tp** [comment]
Yes. I just tried it out with this change. Same behaviour.

**Comment by zoq** [comment]
Okay, thanks for the quick test, I thought maybe the matrix is going to be deconstructed before we call `Train()`.

**Comment by rcurtin** [comment]
valgrind seems to indicate that this is actually coming from the STB loading code:

```
==535800== Invalid read of size 8
==535800==    at 0x483A1BC: memcpy@GLIBC_2.2.5 (vg_replace_strmem.c:1034)
==535800==    by 0x13824B: copy<unsigned char> (arrayops_meat.hpp:36)
==535800==    by 0x13824B: arma::Mat<unsigned char>::Mat(unsigned char*, unsigned long long, unsigned long long, bool, bool) (Mat_meat.hpp:1231)
==535800==    by 0x13289E: bool mlpack::data::Load<unsigned char>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, arma::Mat<unsigned char>&, mlpack::data::ImageInfo&, bool, bool) (load_image_impl.hpp:79)
==535800==    by 0x12C04C: main (cnn_test.cpp:27)
==535800==  Address 0x8f47538 is 8 bytes before a block of size 65,536 alloc'd
==535800==    at 0x4837EC3: memalign (vg_replace_malloc.c:908)
==535800==    by 0x4837FF0: posix_memalign (vg_replace_malloc.c:1072)
==535800==    by 0x13A2BA: unsigned char* arma::memory::acquire<unsigned char>(unsigned long long) (memory.hpp:82)
==535800==    by 0x13D6AD: arma::Mat<unsigned char>::init_cold() (Mat_meat.hpp:220)
==535800==    by 0x1381EC: arma::Mat<unsigned char>::Mat(unsigned char*, unsigned long long, unsigned long long, bool, bool) (Mat_meat.hpp:1229)
==535800==    by 0x13289E: bool mlpack::data::Load<unsigned char>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, arma::Mat<unsigned char>&, mlpack::data::ImageInfo&, bool, bool) (load_image_impl.hpp:79)
==535800==    by 0x12C04C: main (cnn_test.cpp:27)
```

I'm not sure why that is... the code looks correct to me!  I am still digging and trying to figure out what is going on here though.

**Comment by zoq** [comment]
I think I know what happens here, we are loading an image with an alpha channel, but we tell stb to only load 3 channels (r,g,b):

https://github.com/mlpack/mlpack/blob/7f520f032e62c73232cfc097e57824088c839c59/src/mlpack/core/data/load_image_impl.hpp#L60-L61

Here `STBI_rgb` is the number of desired channels to load, which is defined as 3 not 4, looks like there is an `auto` which loads all channels, if we set it to 0.

**Comment by rcurtin** [comment]
Agreed, I think we need to revise the handling of the number of channels to load---maybe if `info.Channels()` is `0`, we auto-detect based on what STB finds, and otherwise we use the given number of channels.

I made a change to load the correct number of channels for the test image I had, but then there was a later segfault in the CNN code.  I'm not sure yet if this is because of the parameters of the CNN or some other issue.

**Comment by adithya-tp** [comment]
The only CNN parameter change that I could suggest, off the top of my head, is changing the output number from 2 to 1 in the linear layer. Maybe that helps.

**Comment by rcurtin** [comment]
I believe the number of input maps needs to be `1`, not `info.Channels()`.

After making that change, I was still encountering memory issues, so I dug deeper and the issue that I am finding is that memory is inadvertently freed, apparently by the `Convolution` layer, during the call to `Forward()`:

```
$ valgrind --leak-check=full --track-origins=yes ./cnn_test 2>&1
...
==792463== Invalid read of size 8
==792463==    at 0x483A024: memcpy@GLIBC_2.2.5 (vg_replace_strmem.c:1034)
==792463==    by 0x131944: copy<double> (arrayops_meat.hpp:36)
==792463==    by 0x131944: arma::Mat<double>::operator=(arma::Mat<double> const&) (Mat_meat.hpp:813)
==792463==    by 0x231EEA: void mlpack::ann::Convolution<mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::FullConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, arma::Mat<double>, arma::Mat<double> >::Gradient<double>(arma::Mat<double> const&&, arma::Mat<double>&&, arma::Mat<double>&&) (convolution_impl.hpp:276)
...
==792463==  Address 0x9161300 is 0 bytes inside a block of size 524,288 free'd
==792463==    at 0x48369AB: free (vg_replace_malloc.c:540)
==792463==    by 0x1300BC: release<double const> (memory.hpp:141)
==792463==    by 0x1300BC: arma::Mat<double>::~Mat() (Mat_meat.hpp:29)
==792463==    by 0x154BF0: double mlpack::ann::FFN<mlpack::ann::NegativeLogLikelihood<arma::Mat<double>, arma::Mat<double> >, mlpack::ann::RandomInitialization>::Eval
uateWithGradient<arma::Mat<double> >(arma::Mat<double> const&, unsigned long, arma::Mat<double>&, unsigned long) (ffn_impl.hpp:325)
...
```

If I remove the `Convolution` layer and replace it with a `Linear` layer (of the correct size), there's no memory leak.  I'm not particularly confident with this part of the code--- @zoq, any ideas?  I can dig deeper, but I think it might take a while before I figured out what was going on.

**Comment by adithya-tp** [comment]
> I believe the number of input maps needs to be 1, not info.Channels().

Could you explain the rationale behind this?

**Comment by rcurtin** [comment]
>> I believe the number of input maps needs to be 1, not info.Channels().
>
> Could you explain the rationale behind this?

The rationale was that I misunderstood the code. :)  Anyway, you're right, it should be `info.Channels()`.

I found the real issue here.  `Convolution` creates an alias to its input during the forward pass, on line 119 of `convolution_impl.hpp`:

```
  inputTemp = arma::cube(const_cast<arma::Mat<eT>&&>(input).memptr(),
      inputWidth, inputHeight, inSize * batchSize, false, false);
```

But `FFN::Forward()` is called using `std::move()`, which means that memory will be freed after the forward pass.  See `ffn_impl.hpp`, line 325:

```
  Forward(std::move(predictors.cols(begin, begin + batchSize - 1)));
```

As a result, when `inputTemp` is used later in `Convolution::Gradient()`, for instance, on line 276, the access is invalid.  The bug can be worked around by changing the creation of `inputTemp` to force a copy (note that this causes slowdown):

```
  inputTemp = arma::cube(const_cast<arma::Mat<eT>&&>(input).memptr(),
      inputWidth, inputHeight, inSize * batchSize, true, false);
```

I think that `Convolution` is not the only layer that does this---it looks like `MaxPooling` and `MeanPooling` might also.

@zoq, @ShikharJ, anyone else--- any thoughts on this?  I didn't write this code so I'm not sure exactly what the best solution is.  Since `FFN::Forward()` simply calls the `ForwardVisitor` with each previous layer's output, it seems like `Convolution` (and other layers) could just take ownership of the input, instead of creating an alias that later becomes unsafe.  Let me know what you think. :+1:

**Comment by unnamed082** [comment]
maybe the error in my   example(https://github.com/mlpack/models/issues/35) was caused by the same memory error, because the error pointed to the system file (I can be wrong)

**Comment by rcurtin** [comment]
@unnamed082 yes, I think this is correct.  A workaround similar to the one I posted for `convolution_impl.hpp` for the other layers in your network could help, I think.

**Comment by walragatver** [comment]
Hi @rcurtin, 
I am trying to figure out the error, I have a doubt, which version are you using for generating the `valgrind` result. Because currently in 3.2.2 line 325 in `ffn_impl.hpp` is slightly different. Also line 276 in `convolution_impl.hpp` is slightly looking different. 
Also I saw 3.1.x branch code it is also looks slightly different.

**Comment by walragatver** [comment]
Hi,
I implemented similar thing what is done `convolution.hpp`. Below is the sample code.
```
# include <bits/stdc++.h>
# include <mlpack/core.hpp>
using namespace std;
# define ll long long int
void make(arma::mat num)
{
	for(auto i = 0 ;i < num.n_cols; i++)
	{
		cout << num(0, i) << " ";
	}
	cout << endl;
}
void call(const arma::mat&& arr)
{
	arma::cube x;
	x = arma::cube(const_cast<arma::mat&&>(arr).memptr(),
               1, 5, 5, false, false);	


	for(auto u : x)
	{
		cout << u << " ";
	}
	cout << endl;
	//make(arr);
}
int main()
{
	int var = 3;
	arma::mat pred = arma::ones(1, 25);
		

	/*vector <ll> arr;
	arr.push_back(8);
	arr.push_back(7);
	arr.push_back(5);
	arr.push_back(3);*/
	call(move(pred));
	for(auto i = 0; i < pred.n_cols; i++)
	{
		cout << pred(0, i) << " ";
	}
	cout << endl;
}
```
You can compile it using `g++ file.cpp -fopenmp` provided you have installed mlpack after building. Looks like the memory doesn't get erased because of an alias. There might be another reason for it.

**Comment by rcurtin** [comment]
Oh, sorry, I thought I was looking on the master branch but I was actually on another checked out branch.  The line in `ffn_impl.hpp` is line 328, and in `convolution_impl.hpp` is line 203.

**Comment by zoq** [comment]
> @zoq, @ShikharJ, anyone else--- any thoughts on this? I didn't write this code so I'm not sure exactly what the best solution is. Since FFN::Forward() simply calls the ForwardVisitor with each previous layer's output, it seems like Convolution (and other layers) could just take ownership of the input, instead of creating an alias that later becomes unsafe. Let me know what you think. 👍

Nice catch, taking ownership of the input is probably the best option to solve the issue, no need to create an extra copy that slows everything down. So I guess we could either open an issue, or I could write a patch for the issue, what do you think?

**Comment by rcurtin** [comment]
While working through the patch this turns out to be more complex than I thought: the `FFN` class will set `outputParameter` to be equal to the previous layer's output, in `Forward()`:

```
  for (size_t i = 1; i < end - begin + 1; ++i)
  {
    boost::apply_visitor(ForwardVisitor(std::move(boost::apply_visitor(
        outputParameterVisitor, network[begin + i - 1])), std::move(
        boost::apply_visitor(outputParameterVisitor, network[begin + i]))),
        network[begin + i]);
  }
```

But then the output is used again in `Backward()`:

```
  for (size_t i = 2; i < network.size(); ++i)
  {
    boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
        outputParameterVisitor, network[network.size() - i])), std::move(
        boost::apply_visitor(deltaVisitor, network[network.size() - i + 1])),
        std::move(boost::apply_visitor(deltaVisitor,
        network[network.size() - i]))), network[network.size() - i]);
  }
```

So if a layer that takes ownership of the input is used as a layer that's not the first layer, this will cause problems.

It seems like the only way around for now would be to make a copy, and in the longer-term a more significant redesign might be needed; I'm not sure.  Let me know what you think the best way forward is...

**Comment by zoq** [comment]
I'll take a closer look into the issue tomorrow, but making a copy is a solution at least for now.

**Comment by rcurtin** [comment]
I've already got most of a PR (it's quite simple); I'll try and open it tonight.

**Comment by rcurtin** [comment]
PR #2234 should fix this, although perhaps it is not the best possible solution. :+1:

**Comment by adithya-tp** [comment]
Since @rcurtin has resolved the issue, shall I close this?

**Comment by rcurtin** [comment]
#2234 isn't merged yet, so let's wait on that first. :+1:
